wcfsetup/install/files/lib/acp/form/LanguageAddForm.class.php

```diff
@@ -103,149 +83,107 @@ public function readParameters()
-        if (!isset($this->locales[$this->locale])) {
-            throw new UserInputException('locale');
-        }
-    }
+        $this->languages = LanguageFactory::getInstance()->getLanguages();
 
-    /**
-     * Validates the language code.
-     */
-    protected function validateLanguageCode()
-    {
-        if (empty($this->languageCode)) {
-            throw new UserInputException('languageCode');
-        }
-        if (LanguageFactory::getInstance()->getLanguageByCode($this->languageCode)) {
-            throw new UserInputException('languageCode', 'notUnique');
-        }
+        $locales = [
+            '' => WCF::getLanguage()->get('wcf.global.noSelection'),
+            ...$this->locales
+        ];
+
+        $this->form->appendChildren([
+            FormContainer::create('data')
+                ->appendChildren([
+                    TextFormField::create('languageName')
+                        ->label('wcf.global.name')
+                        ->description('wcf.acp.language.name.description')
+                        ->maximumLength(255)
+                        ->required(),
+                    TextFormField::create('languageCode')
+                        ->label('wcf.acp.language.code')
+                        ->description('wcf.acp.language.code.description')
+                        ->maximumLength(20)
+                        ->required()
+                        ->addValidator(new FormFieldValidator('unique', function (TextFormField $formField) {
+                            if ($formField->getValidationErrors() !== []) {
+                                return;
+                            }
+
+                            if (
+                                $this->formObject instanceof Language
+                                && \mb_strtolower($this->formObject->languageCode) === $formField->getValue()
+                            ) {
+                                return;
+                            }
+
+                            if (LanguageFactory::getInstance()->getLanguageByCode($formField->getValue())) {
+                                $formField->addValidationError(new FormFieldValidationError(
+                                    'notUnique',
+                                    'wcf.acp.language.add.languageCode.error.notUnique'
+                                ));
+                            }
+                        })),
+                    TextFormField::create('countryCode')
+                        ->label('wcf.acp.language.countryCode')
+                        ->description('wcf.acp.language.countryCode.description')
+                        ->maximumLength(10)
+                        ->required(),
+                    SingleSelectionFormField::create('locale')
+                        ->label('wcf.acp.language.locale')
+                        ->description('wcf.acp.language.locale.description')
+                        ->options($locales)
+                        ->required(),
```

Comment:
The `required()` should not be there: Not selecting anything is fine.

```suggestion
```

Reply:
It was required before, are you sure?
However this case would benefit from https://github.com/WoltLab/WCF/pull/5159/commits/e821b790d581d029384e07d4ab4c36c11febba3e

Reply:
Yes, I'm sure about this: The `required()` implies that the user has to make a selection, when actually “No Selection” is fine. Completely deleting the `<select>` from the HTML will result in a `null` return even if the field is `required()`, so this is definitely is a bug there. It appears that support for `required()` is not implemented at all. I'll file an issue for this.

I also agree that this would likely benefit from the commit you linked, however that should be reviewed and included separately. I have some concerns about the added API surface that we'd need to discuss, instead of making that a separate flag it should likely rely on `required()` to fix the behavior of `required()`. 6.0 is the right version to make this fix, as it can be slightly breaking. Would you be interested in making a PR for this?